<a href="https://colab.research.google.com/github/mayujie/MSC_text_mining/blob/master/RD_Embedding_testing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# [Amazon product data](http://jmcauley.ucsd.edu/data/amazon/index_2014.html)
# [Amazon product data Julian McAuley, UCSD](http://jmcauley.ucsd.edu/data/amazon/)
Files

"Small" subsets for experimentation

https://zhuanlan.zhihu.com/p/137537499

http://baijiahao.baidu.com/s?id=1665370803562402243

http://www.xiaomaidong.com/?p=1249

https://towardsdatascience.com/double-your-google-colab-ram-in-10-seconds-using-these-10-characters-efa636e646ff


In [ ]:
from google.colab import drive
drive.mount('/content/drive/')

Mounted at /content/drive/


In [ ]:
root_path = '/content/drive/My Drive/Colab/MSC_RD'
root_path

'/content/drive/My Drive/Colab/MSC_RD'

## **Read from cleaned text file**

In [ ]:
### download cleaned text data
# !gdown https://drive.google.com/uc?id=1Kb8ApeUYUDMOO39utV-dQ00PZ0s8yOAb 
# !unzip cleaned_reviews_text.zip
!unzip /content/drive/MyDrive/Colab/MSC_RD/cleaned_reviews_text.zip
!rm /content/cleaned_reviews_text.zip
!ls -lh *.txt

Archive:  /content/drive/MyDrive/Colab/MSC_RD/cleaned_reviews_text.zip
  inflating: cleaned_reviews_text.txt  
rm: cannot remove '/content/cleaned_reviews_text.zip': No such file or directory
-rw-r--r-- 1 root root 241M Nov 13 13:20 cleaned_reviews_text.txt


In [ ]:
cleaned_text_ls = []
with open('/content/cleaned_reviews_text.txt', 'r') as file:
  for line in file:
    cleaned_text_ls.append(line)

In [ ]:
print(len(cleaned_text_ls))
cleaned_text_ls[:5]

484826


['big expectations love english tv <COMMA> particular investigative detective stuff guy really boring <PERIOD> not appeal <PERIOD>\n',
 'highly recommend series <PERIOD> must anyone yearning watch grown television <PERIOD> complex characters plots keep one totally involved <PERIOD> thank amazin prime <PERIOD>\n',
 'one real snoozer <PERIOD> not believe anything read hear <COMMA> awful <PERIOD> no idea title means <PERIOD> neither <PERIOD>\n',
 'mysteries interesting <PERIOD> tension robson tall blond good not always believable <PERIOD> often seemed uncomfortable <PERIOD>\n',
 'show always excellent <COMMA> far british crime mystery one best ever made <PERIOD> stories well done acting top notch interesting twists realistic brutal storylines <PERIOD> show pulls no punches enters twisted minds criminals profiler psychiatrist helps northern english city police force <PERIOD> show looks like shot manchester called another name show <PERIOD> one episode not disc excellent prayer bone seperat

## remove newline '\n' from each row

In [ ]:
cleaned_text_ls = map(lambda s: s.strip('\n'), cleaned_text_ls)
cleaned_text_ls = list(cleaned_text_ls)
cleaned_text_ls[:5]

['big expectations love english tv <COMMA> particular investigative detective stuff guy really boring <PERIOD> not appeal <PERIOD>',
 'highly recommend series <PERIOD> must anyone yearning watch grown television <PERIOD> complex characters plots keep one totally involved <PERIOD> thank amazin prime <PERIOD>',
 'one real snoozer <PERIOD> not believe anything read hear <COMMA> awful <PERIOD> no idea title means <PERIOD> neither <PERIOD>',
 'mysteries interesting <PERIOD> tension robson tall blond good not always believable <PERIOD> often seemed uncomfortable <PERIOD>',
 'show always excellent <COMMA> far british crime mystery one best ever made <PERIOD> stories well done acting top notch interesting twists realistic brutal storylines <PERIOD> show pulls no punches enters twisted minds criminals profiler psychiatrist helps northern english city police force <PERIOD> show looks like shot manchester called another name show <PERIOD> one episode not disc excellent prayer bone seperate disc <

In [ ]:
# ## read data into dataframe
# import pandas as pd
# df_review = pd.DataFrame(cleaned_text_ls) 
# df_review

In [ ]:
# cleaned_text_ls[9568]

In [ ]:
# def find_noise(reg, datalist):
#   count = 0
#   found_list = []

#   for i, row in enumerate(datalist):
#     found = re.findall(reg, row)

#     if found != []:
#       count += 1
#       # print(found[0])
#       # found = list(filter(None, found[0]))
#       # print(found)
#       found_list.append([i ,found])
  
#   return count, found_list
#   # return found_list

In [ ]:
# import re
# ## check tokens for each row
# count_key, found_key = find_noise(r"<\w+>", cleaned_text_ls) ## "<\w+>" " <UNK> "
# print(count_key, len(found_key))
# found_key[:10]

In [ ]:
# ss = [i[0] for i in found_key]
# print(ss)

In [ ]:
# ss = []
# for i in found_key:
#   for j in i[1]: 
#     ss.append(j)

# from collections import Counter 
# print(len(Counter(ss).most_common()))
# Counter(ss).most_common() # 12, 14 tokens with each amount

**empty reviews: [(' \<UNK\> ', 138)]**

In [ ]:
# reviews_text_merge = " ".join(cleaned_text_ls)
# print(len(reviews_text_merge)) ## 252078789
# reviews_text_merge[:1000]

In [ ]:
# with open("reviews_text_merge.txt", "w") as file:
#     file.write("%s" % reviews_text_merge)

In [ ]:
# # Compress review text merge into zip file
# !zip "{root_path}"/reviews_text_merge.zip reviews_text_merge.txt

In [ ]:
# # https://drive.google.com/file/d/1Wrw1Wk2w6pkdfdQ3asOTX1_s_7djo1yk/view?usp=sharing
# !unzip "{root_path}"/reviews_text_merge.zip

In [ ]:
# with open('/content/reviews_text_merge.txt') as file:
#     text = file.read()

In [ ]:
# # words = text.split()
# words = " ".join(cleaned_text_ls).split()

In [ ]:
# from collections import Counter 

# word_counts = Counter(words)
# print('Total number of words in cleaned corpus:', len(words))
# print('Number of unique words in cleaned corpus:', len(Counter(words)), len(Counter(words).most_common()))

Total number of words in cleaned corpus: 36436399

Number of unique words in cleaned corpus: 55984 55984

In [ ]:
# word_counts.most_common()[44187-10:]

In [ ]:
# vocab_to_int, int_to_vocab = create_lookup_tables(words)

In [ ]:
# int_words = [vocab_to_int[word] for word in words]
# print(words[:30])
# print(int_words[:30])

['big', 'expectations', 'love', 'english', 'tv', '<COMMA>', 'particular', 'investigative', 'detective', 'stuff', 'guy', 'really', 'boring', '<PERIOD>', 'not', 'appeal', '<PERIOD>', 'highly', 'recommend', 'series', '<PERIOD>', 'must', 'anyone', 'yearning', 'watch', 'grown', 'television', '<PERIOD>', 'complex', 'characters']

[101, 1803, 20, 1865, 791, 1, 749, 19071, 4452, 169, 785, 16, 1481, 0, 2, 2068, 0, 198, 90, 318, 0, 226, 311, 8751, 409, 1978, 2657, 0, 1951, 577]

## create word_counts dictionary for all words

In [ ]:
%whos

Variable          Type             Data/Info
--------------------------------------------
cleaned_text_ls   list             n=484826
drive             module           <module 'google.colab.dri<...>s/google/colab/drive.py'>
file              TextIOWrapper    <_io.TextIOWrapper name='<...>ode='r' encoding='UTF-8'>
line              str              nice head lamp <EXCLAMATI<...>flashing modes <PERIOD>\n
root_path         str              /content/drive/My Drive/Colab/MSC_RD


In [ ]:
words = []
for row in cleaned_text_ls:
  for i in row.split():
    # print(i)
    words.append(i)

In [ ]:
from collections import Counter 

print('Total number of words in cleaned corpus:', len(words))
print('Number of unique words in cleaned corpus:', len(Counter(words)), len(Counter(words).most_common()))

Total number of words in cleaned corpus: 36436399
Number of unique words in cleaned corpus: 55984 55984


### **For 8 classes data (Current):**
Length of reviews: **484826**

Total number of words in cleaned corpus: **36436399**

Number of unique words in cleaned corpus: **55984** **55984**

### For 6 classes data:

Length of reviews: **199096**

Total number of words in cleaned corpus: 18241862

Number of unique words in cleaned corpus: 44187 44187

### For 10 classes data:

Length of reviews: **2360871**

Total number of words in cleaned corpus: 145318404

Number of unique words in cleaned corpus: 110454 110454

In [ ]:
# Remove all words with  5 or fewer occurences
from collections import Counter 

def create_lookup_tables(words):
  word_counts = Counter(words)
  # print(word_counts.most_common(10))
  # print(word_counts.most_common())
  ## sorting the words from most to least frequent in text occurrence, descending
  sorted_vocab = sorted(word_counts, key=word_counts.get, reverse=True)
  # print(len(sorted_vocab)) # 110454
  ## create int_to_vocab dictionaries
  int_to_vocab = {ii: word for ii, word in enumerate(sorted_vocab)}
  vocab_to_int = {word: ii for ii, word in int_to_vocab.items()}

  return vocab_to_int, int_to_vocab

[('\<PERIOD\>', 1671352),
 ('\<COMMA\>', 1266882),
 ('not', 341506),
 ('album', 140815),
 ('one', 140781),
 ('like', 127551),
 ('so', 120402),
 ('\<EXCLAMATION\>', 95972),
 ('just', 95320),
 ('good', 89656)]

12 Tokens for 6 classes data

[('\<PERIOD\>', 1671352),
 ('\<COMMA\>', 1266882),
 ('\<EXCLAMATION\>', 95972),
 ('\<QUESTION\>', 38148),
 ('\<PRICE\>', 9371),
 ('\<HUNDREDPERCENTAGE\>', 1098),
 ('\<NOTCONTRACT\>', 638),
 ('\<MENTION\>', 197),
 ('\<AMAZONURL\>', 185),
 ('\<URL\>', 136),
 ('\<UNK\>', 54),
 ('\<YOUTUBEURL\>', 8)]

 **14 Tokens for 8 classes data (Current)**

 [('\<PERIOD\>', 3476956),
 ('\<COMMA\>', 2399259),
 ('\<EXCLAMATION\>', 211162),
 ('\<QUESTION\>', 64627),
 ('\<PRICE\>', 29257),
 ('\<HUNDREDPERCENTAGE\>', 4389),
 ('\<NOTCONTRACT\>', 986),
 ('\<AMAZONURL\>', 506),
 ('\<MENTION\>', 288),
 ('\<URL\>', 246),
 ('\<UNK\>', 138),
 ('\<YOUTUBEURL\>', 25),
 ('\<GOOGLEURL\>', 13),
 ('\<STAR\>', 9)]

## lookup table (word to int and vice versa)

In [ ]:
%whos

Variable               Type             Data/Info
-------------------------------------------------
Counter                type             <class 'collections.Counter'>
cleaned_text_ls        list             n=484826
create_lookup_tables   function         <function create_lookup_tables at 0x7fbc82550c80>
drive                  module           <module 'google.colab.dri<...>s/google/colab/drive.py'>
file                   TextIOWrapper    <_io.TextIOWrapper name='<...>ode='r' encoding='UTF-8'>
i                      str              <PERIOD>
line                   str              nice head lamp <EXCLAMATI<...>flashing modes <PERIOD>\n
root_path              str              /content/drive/My Drive/Colab/MSC_RD
row                    str              nice head lamp <EXCLAMATI<...>> flashing modes <PERIOD>
words                  list             n=36436399


In [ ]:
vocab_to_int, int_to_vocab = create_lookup_tables(words)

In [ ]:
## words into int numbers
int_words = [vocab_to_int[word] for word in words]

print(len(int_words))
print(words[:20])
print(cleaned_text_ls[0])
print(int_words[:20])

36436399
['big', 'expectations', 'love', 'english', 'tv', '<COMMA>', 'particular', 'investigative', 'detective', 'stuff', 'guy', 'really', 'boring', '<PERIOD>', 'not', 'appeal', '<PERIOD>', 'highly', 'recommend', 'series']
big expectations love english tv <COMMA> particular investigative detective stuff guy really boring <PERIOD> not appeal <PERIOD>
[101, 1803, 20, 1865, 791, 1, 749, 19071, 4452, 169, 785, 16, 1481, 0, 2, 2068, 0, 198, 90, 318]


In [ ]:
## (just for check, optional)
# int_words = []
# for row in cleaned_text_ls:
#   int_words.append([vocab_to_int[word] for word in row.split()])

# print(len(int_words))
# print(words[:20])
# print(cleaned_text_ls[0])
# print(int_words[0])

In [ ]:
# cleaned_text_ls = [row.split() for row in cleaned_text_ls]
# # print(cleaned_text_ls[:2])
# len(cleaned_text_ls) #484826

In [ ]:
# ## convert each review to word tokens array
# for i, row in enumerate(cleaned_text_ls):
#   cleaned_text_ls[i] = [word for word in row.split()]

# len(cleaned_text_ls) # 484826

# Subsampling (optional) 
I tried in this case, but seems not good. Because i've done stopwords removal before, and then I perform subsampling it leads to appearing a lot of review with lenght 0 or 1 word.

In [ ]:
print('total number of word ', len(words))
print(len(Counter(words)), len(Counter(int_words)))

total number of word  36436399
55984 55984


In [ ]:
print(cleaned_text_ls[0])
print(words[:20])
print(int_words[:20])

big expectations love english tv <COMMA> particular investigative detective stuff guy really boring <PERIOD> not appeal <PERIOD>
['big', 'expectations', 'love', 'english', 'tv', '<COMMA>', 'particular', 'investigative', 'detective', 'stuff', 'guy', 'really', 'boring', '<PERIOD>', 'not', 'appeal', '<PERIOD>', 'highly', 'recommend', 'series']
[101, 1803, 20, 1865, 791, 1, 749, 19071, 4452, 169, 785, 16, 1481, 0, 2, 2068, 0, 198, 90, 318]


In [ ]:
# from collections import Counter
# import random
# import numpy as np

# threshold = 1e-5
# word_counts = Counter([vocab_to_int[word] for word in words])

# # # print 1st key value pair in list
# # print(list(word_counts.items())[0]) #  dictionary of int_words, how many times they appear
# total_count = len([vocab_to_int[word] for word in words])
# print('total_count number of word: ', total_count)

# # calculate the frequency of occurrence for each word in our vocab
# freqs = {word: count/total_count for word, count in word_counts.items()}
# p_drop = {word: 1 - np.sqrt(threshold/freqs[word]) for word in word_counts}

# # discard some frequent words, according to the subsampling equation
# # create a new list of words for training
# # generate a random value between 0 and 1, then check if that value < 1 - drop_p for that word , i want to keep this word with a prob of 1 - drop_p
# # if drop_p is 0.98, then keep word prob is 1 - 0.98 = 0.02, which will be 0.02. if generate a value < 0.02, then it will keep this word in list

# # most of 0 and 1 are gone, subsampling for "words" variable
# train_words = [word for word in [vocab_to_int[word] for word in words] if random.random() < (1 - p_drop[word])]

# print(train_words[:30])
# print('len of train word', len(train_words))

In [ ]:
## method of subsampling for each row 
# train_words = []
# for row in int_words:
#   train_words.append([word for word in row if random.random() < (1 - p_drop[word])])

[20, 749, 19071, 4452, 785, 1481, 2068, 90, 318, 311, 8751, 409, 1978, 2657, 1951, 2300, 34267, 32079, 7315, 2702, 46542, 15474, 4064, 2423, 1728, 1224, 2996, 6078, 7786, 3518]

len of train word 9193920

In [ ]:
# # len(cleaned_text_ls), len(train_words) # (484826, 484826)
# idx = 20
# print(cleaned_text_ls[0])
# print(int_words[:idx])
# print(words[:idx])
# print(train_words[:idx])
# print([int_to_vocab[i] for i in train_words[:idx]])

['comedy', 'matter', 'taste', '<PERIOD>', 'guy', 'little', 'funny', 'others', 'far', 'funnier', 'guy']

[2628, 578, 30, 0, 785, 21, 902, 241, 100, 12538, 785]

[785, 12538, 785]

['guy', 'funnier', 'guy']

In [ ]:
# ls = []
# for row in train_words:
#   for i in row:
#     ls.append(i)

# len(ls) ## 9188684 words left after subsampling

In [ ]:
# ls_word = [int_to_vocab[i] for i in ls] ## convert to word

In [ ]:
# [i for i,j in enumerate(train_words) if (len(j)==0)] ## 5607 empty row appears
# [i for i,j in enumerate(train_words) if (len(j)==1)] ## 15105 rows with only one word
# [i for i,j in enumerate(train_words) if (len(j)==2)] ## 24584 rows with only two word

# Making batches of data

window size

http://jalammar.github.io/illustrated-word2vec/#:~:text=The%20Gensim%20default%20window%20size,factor%20of%20the%20training%20process.

https://stackoverflow.com/questions/22272370/word2vec-effect-of-window-size-used/30447723

https://stackoverflow.com/questions/59872029/word2vec-window-size-at-sentence-boundaries

In [ ]:
%whos

Variable               Type             Data/Info
-------------------------------------------------
Counter                type             <class 'collections.Counter'>
cleaned_text_ls        list             n=484826
create_lookup_tables   function         <function create_lookup_tables at 0x7fbc82550c80>
drive                  module           <module 'google.colab.dri<...>s/google/colab/drive.py'>
file                   TextIOWrapper    <_io.TextIOWrapper name='<...>ode='r' encoding='UTF-8'>
i                      str              <PERIOD>
int_to_vocab           dict             n=55984
int_words              list             n=36436399
line                   str              nice head lamp <EXCLAMATI<...>flashing modes <PERIOD>\n
root_path              str              /content/drive/My Drive/Colab/MSC_RD
row                    str              nice head lamp <EXCLAMATI<...>> flashing modes <PERIOD>
vocab_to_int           dict             n=55984
words                  list       

In [ ]:
import numpy as np

def get_target(words, idx, window_size=5):
  ''' Get a list of words in a window around an index. '''
  
  # R is random integer in the range of 1:C, R is the size
  R = np.random.randint(1, window_size+1)
  print("R: ", R, " len of input sentence ", len(words))
  
  # define the start and stop indices of context window
  # start will be a range of words in the past, index of word interest - R, only if it not gives negative idx
  # if it gives a negative value, then just set idx to the start of list of words idx 0
  start = idx - R if (idx - R) > 0 else 0
  
  # stop idx is where future words end
  stop = idx + R
  
  # finally we don't want my target context to include the word at the past (word interest) in idx
  target_words = words[start:idx] + words[idx+1:stop+1]

  return list(target_words)

In [ ]:
idx = 6 # word index of interest
print('Input: ', int_words[:10])

target = get_target(int_words[:10], idx, window_size=5)
print('Target: ', target)  # you should get some indices around the idx

Input:  [101, 1803, 20, 1865, 791, 1, 749, 19071, 4452, 169]
R:  3  len of input sentence  10
Target:  [1865, 791, 1, 19071, 4452, 169]


# Generating Batches